### PR TITLE
Introduce test task option to fail if there is no test

### LINKED ISSUE #7452: Introduce an easy fail if no tests config option
There should be an easy config option to mark that the build should fail if no tests were executed. In Maven, it's very easy to achieve by setting the surefire plugin option like so: `<failIfNoTests>true</failIfNoTests>`. There should be an easy option to set it up in gradle.

### Expected Behavior
It would be good to have an easily available config option, like:

```[groovy]
test {
	failIfNoTests()
}
```

### Current Behavior
Right now, the following workaround is possible:
```[groovy]
test {
	afterSuite { desc, result ->
		if (!desc.parent) {
			if (result.testCount == 0) {
				throw new IllegalStateException("No tests were found. Failing the build")
			}
		}
	}
}
```

### PR DESCRIPTION
**Pull request by FrauBoes**
work in progress

Fixes #7452 

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation

### CODE REVIEW

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TestTaskNoTestsFailTest.groovy [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.executer.UnexpectedBuildFailure
+
+class TestTaskNoTestsFailTest extends AbstractIntegrationSpec {
+
+    def "test --failIfNoTest fails if there is no test"() {
+        buildFile << "apply plugin: 'java'"
+
+        file("src/test/java/NotATest.java") << """
+            public class NotATest {}
```
Comment (eskatos) [inline subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TestTaskNoTestsFailTest.groovy:29]: We need also coverage for when there are tests using one testing framework but the test task is configured to use another testing framework.

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TestTaskNoTestsFailTest.groovy [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+        when:
+        executer.expectDeprecationWarning("There is no test to run. You can use 'test --failIfNoTest' to fail in this case.")
+        succeeds("test")
+
+        then:
+        noExceptionThrown()
+    }
+
+    // These don't work due to the @SkipWhenEmpty annotation on Test::getCandidateClassFiles
```
Comment (eskatos) [inline subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TestTaskNoTestsFailTest.groovy:57]: This is fine, the test task should still be skipped if it has no "candidate class files".

**Review by eskatos — CHANGES_REQUESTED** [review]
What we want Today is
- by default, if there are no tests emit a deprecation warning that this behavior will change in 9.0 with an advice to use `failIfNoTest` if needed
- if `failIfNoTest` is set to `true` then we fail if there are no tests
- if `failIfNoTest` is set to `false` then do not fail, do not warn

What we want in Gradle 9.0
- by default, `failIfNoTest` is set to `true`, fail if there are no tests
- if `failIfNoTest` is set to `false` then do not fail, do not warn

In order to do that we should Today:
- make `failIfNoTest` `@Optional` with no default value

In Gradle 9.0 we will:
- remove `@Optional` on `failIfNoTest`
- set the conventional value of `failIfNoTest` to `true`

**Comment by FrauBoes** [comment]
As @Paul lined out, we need three cases to facilitate the deprecation of the current behaviour:
1. `--failIfNoTest` option not present
2. `--failIfNoTest=true`
3. `--failIfNoTest=false`

The intuitive approach is to use a `Property<Boolean>` here. Unfortunately, the implementation class `BooleanOptionElement` is designed as a flag, meaning the presence or absence of the option signifies true or false, no value can be passed to the flag, which gives us only two cases. I therefore used a `Property<String>` and limit the possible values to "true" and "false" using the `OptionValues` annotation. 

An alternative approach is what's done for build command line options, where the implementation class `BooleanBuildOption` auto-creates the boolean counter flag, e.g. `--daemon` / `--no-daemon`. This would be a bit more involved and I don't think `--no-failIfNoTest` works well semantically.

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java [hunk 3]
```diff
@@ -566,6 +575,22 @@ void setFailFast(boolean failFast) {
         this.failFast = failFast;
     }
 
+    /**
+     * Indicates if this task will succeed when there is no test to run.
+     *
+     * @return whether this task will succeed when there is no test to run
+     */
+    Property<Boolean> getSuccessWithoutTest() {
```
Comment (eskatos) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java:583]: :x: This needs to be marked as `@Internal` otherwise other descendent of `AbstractTestTask` than `Test` will fail task properties validation

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java [hunk 4]
```diff
@@ -481,8 +484,14 @@ public void execute(DeleteSpec spec) {
     private void handleCollectedResults(TestCountLogger testCountLogger) {
         if (testCountLogger.hadFailures()) {
             handleTestFailures();
-        } else if (testCountLogger.getTotalTests() == 0 && shouldFailOnNoMatchingTests()) {
-            throw new TestExecutionException(createNoMatchingTestErrorMessage());
+        } else if (testCountLogger.getTotalTests() == 0) {
+            if (shouldFailOnNoMatchingTests()) {
+                throw new TestExecutionException(createNoMatchingTestErrorMessage());
+            } else if (successWithoutTest.getOrNull() == null) {
+                getLogger().warn("There is no test to run. In 9.0, the behaviour will change to fail in this case. Use '--success-without-test' or '--no-success-without-test' to set the behaviour.");
```
Comment (eskatos) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java:491]: :x: This should be a documented deprecation warning

Something like:
```java
DeprecationLogger.deprecateBehaviour("There is no test to run.")
    .withAdvice("Set Test.successWithoutTest to true if you want the task to succeed when there is no test to run.")
    .willBecomeAnErrorInGradle9()
    .withUpgradeGuideSection(8, "TODO")
    .nagUser();
```

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TaskBooleanOptionIntegrationTest.groovy [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+                public void setMyBooleanObjectOption(Boolean myBooleanObjectOption) {
+                    this.myBooleanObjectOption = myBooleanObjectOption;
+                }
+
+                @Option(option = "myBooleanPropertyOption", description = "Configures Property<Boolean> option 'myBooleanPropertyOption'")
+                public void setMyBooleanPropertyOption(Boolean myBooleanPropertyOption) {
+                    this.myBooleanPropertyOption.set(myBooleanPropertyOption);
+                }
+
+                @OptionValues("myBooleanObjectOption")
```
Comment (eskatos) [inline subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TaskBooleanOptionIntegrationTest.groovy:205]: :question: What is this used for in this test?

File: subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/options/OptionReaderTest.groovy [hunk 6]
```diff
@@ -169,6 +185,15 @@ class OptionReaderTest extends Specification {
...
     }
 
+    def "fails when option name starts with 'no-'"() {
+        when:
+        reader.getOptions(new TestClass6());
+        then:
+        def e = thrown(OptionValidationException)
+        println(e.message)
+        e.message == "Illegal option name 'no-optionName' in class 'org.gradle.api.internal.tasks.options.OptionReaderTest\$TestClass6'. Option name must not start with 'no-'."
+    }
```
Comment (eskatos) [inline subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/options/OptionReaderTest.groovy:195]: :x: This is a breaking change for builds that have task which declare a boolean option whose name starts with `no-`.

What we should do instead is only register disabling command line options iff they don't conflict with a declared option. See `OptionDescriptor.isClashing()`.

**Review by eskatos — CHANGES_REQUESTED** [review]

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TaskBooleanOptionIntegrationTest.groovy (hunk 5 above)
Comment (FrauBoes) [reply subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TaskBooleanOptionIntegrationTest.groovy:205]: Good catch, a leftover from another test I used. Removed.

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java [hunk 7]
```diff
@@ -481,8 +485,18 @@ public void execute(DeleteSpec spec) {
...
-        } else if (testCountLogger.getTotalTests() == 0 && shouldFailOnNoMatchingTests()) {
-            throw new TestExecutionException(createNoMatchingTestErrorMessage());
+        } else if (testCountLogger.getTotalTests() == 0) {
+            if (shouldFailOnNoMatchingTests()) {
+                throw new TestExecutionException(createNoMatchingTestErrorMessage());
+            } else if (successWithoutTest.getOrNull() == null) {
+                DeprecationLogger.deprecateBehaviour("There is no test to run.")
+                    .withAdvice("Set Test.successWithoutTest to true if you want the task to succeed when there is no test to run.")
+                    .willBecomeAnErrorInGradle9()
+                    .withUpgradeGuideSection(8, "TODO")
```
Comment (eskatos) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java:495]: :x: You need to add a section to the upgrade guide and link it here
:lipstick: Also extracting the deprecation logging in a dedicated method would make this clearer.

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TestTaskSuccessWithoutTestIntegrationTest.groovy [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+    def "test succeeds with warning if there is no test"() {
+        buildFile << "apply plugin: 'java'"
+
+        file("src/test/java/NotATest.java") << """
+            public class NotATest {}
+        """
+
+        when:
+        executer.expectDeprecationWarning("There is no test to run. This behavior has been deprecated. This will fail with an error in Gradle 9.0. Set Test.successWithoutTest to true if you want the task to succeed when there is no test to run. Consult the upgrading guide for further information: https://docs.gradle.org/")
```
Comment (eskatos) [inline subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/TestTaskSuccessWithoutTestIntegrationTest.groovy:33]: :x: This should use `executer.expectDocumentedDeprecationWarning()` instead. Same of all other instances.

File: subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/options/OptionReaderTest.groovy [hunk 9]
```diff
@@ -169,6 +185,15 @@ class OptionReaderTest extends Specification {
         e.message == "Option 'stringValue' on method cannot take multiple parameters in class 'org.gradle.api.internal.tasks.options.OptionReaderTest\$TestClass4#setStrings'."
     }
 
+    def "fails when option name starts with 'no-'"() {
```
Comment (eskatos) [inline subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/options/OptionReaderTest.groovy:188]: :x: This test should be removed now. A new one that assert that we don't add the disabling option dynamically if there's already an option with that name should be added.

File: subprojects/core/src/main/java/org/gradle/execution/commandline/CommandLineTaskConfigurer.java [hunk 10]
```diff
@@ -94,4 +102,8 @@ private List<String> configureTasksNow(Collection<Task> tasks, List<String> argu
         }
         return remainingArguments;
     }
+
+    private static void allowOnlyOneOption(OptionDescriptor optionDescriptor, CommandLineParser parser) {
+        parser.allowOneOf(optionDescriptor.getName(), optionDescriptor.getName().substring(3));
```
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/execution/commandline/CommandLineTaskConfigurer.java:107]: :art: I think this should be reworked. The knowledge of doing substrings is spread apart, this should be in a central place or encapsulated.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/options/OptionReader.java [hunk 11]
```diff
@@ -185,7 +193,11 @@ private List<OptionElement> getFieldAnnotations(Class<?> type) {
         for (Field field : type.getDeclaredFields()) {
             Option option = findOption(field);
             if (option != null) {
-                fieldOptionElements.add(FieldOptionElement.create(option, field, optionValueNotationParserFactory));
+                OptionElement optionElement = FieldOptionElement.create(option, field, optionValueNotationParserFactory);
+                fieldOptionElements.add(optionElement);
+                if (optionElement.getOptionType().equals(Void.TYPE)) {
```
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/options/OptionReader.java:198]: :x: This test is looking at the option argument type. There are more options than `BooleanOptionElement` that have `void` argument type (e.g. `--rerun`). This might be the cause of the failures in `CommandLineTaskConfigurerSpec`. It would make more sense to test if `optionElement instanceOf BooleanOptionElement`.

But, that being said, I would say that the creation of the disabling options could be entirely moved out of `OptionReader` into `CommandLineTaskConfigurer`. Then the handling of clashing options and dis-allowance of mutually exclusive options could be simplified.

**Review by eskatos — CHANGES_REQUESTED** [review]

File: subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/options/OptionReaderTest.groovy (hunk 9 above)
Comment (eskatos) [reply subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/options/OptionReaderTest.groovy:188]: The new test mentioned here is still missing. Or did I miss it?

Comment (FrauBoes) [reply subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/options/OptionReaderTest.groovy]: That's added now, I left it for after the implementation changes :)

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java [hunk 12]
```diff
@@ -120,6 +123,7 @@ public AbstractTestTask() {
         testOutputListenerBroadcaster = listenerManager.createAnonymousBroadcaster(TestOutputListener.class);
         testListenerBroadcaster = listenerManager.createAnonymousBroadcaster(TestListener.class);
         binaryResultsDirectory = getProject().getObjects().directoryProperty();
+        successWithoutTest = getProject().getObjects().property(Boolean.class).convention((Boolean)null);
```
Comment (eskatos) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java:126]: :x: No need for a conventional value
```suggestion
        successWithoutTest = getProject().getObjects().property(Boolean.class);
```

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/options/OptionReader.java [hunk 13]
```diff
@@ -82,14 +83,14 @@ public List<OptionDescriptor> getOptions(Object target, boolean validOnly) {
             JavaMethod<Object, Collection> optionValueMethod = cachedOptionValueMethods.get(optionElement);
             options.put(optionElement.getOptionName(), new InstanceOptionDescriptor(target, optionElement, optionValueMethod));
         }
+        options.putAll(CommandLineTaskConfigurer.getOppositeOptions(options.values(), target));
```
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/options/OptionReader.java:86]: :x: This adds a cyclic dependency between `OptionReader` and `CommandLineTaskConfigurer`!

It would be best to introduce a new type between `CommandLineTaskConfigurer` and `OptionReader`, keep `OptionReader` about just reading options from task classes and `CommandLineTaskConfigurer` about just configuring the task from the command line. The newly introduced type would be responsible for handling both built-in task options (`--rerun`) and boolean opposite options, it would be using `OptionReader` under the hood and be used by both `CommandLineTaskConfigurer` and the `Help` task`.

**Review by eskatos — CHANGES_REQUESTED** [review]

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/options/OptionReader.java (hunk 13 above)
Comment (FrauBoes) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/options/OptionReader.java]: Argh, true! Thanks for the clarification, I updated this accordingly.

File: subprojects/core/src/main/java/org/gradle/execution/commandline/CommandLineTaskConfigurer.java [hunk 14]
```diff
@@ -57,7 +60,7 @@ private List<String> configureTasksNow(Collection<Task> tasks, List<String> argu
         List<String> remainingArguments = null;
         for (Task task : tasks) {
             CommandLineParser parser = new CommandLineParser();
-            Map<Boolean, List<OptionDescriptor>> allCommandLineOptions = optionReader.getOptions(task, false).stream().collect(Collectors.groupingBy(OptionDescriptor::isClashing));
+            Map<Boolean, List<OptionDescriptor>> allCommandLineOptions = TaskOptionSupplier.get(task, optionReader, false).stream().collect(Collectors.groupingBy(OptionDescriptor::isClashing));
             List<OptionDescriptor> validCommandLineOptions = allCommandLineOptions.getOrDefault(false, Collections.emptyList());
             List<OptionDescriptor> clashingOptions = allCommandLineOptions.getOrDefault(true, Collections.emptyList());
             clashingOptions.forEach(it -> LOGGER.warn("Built-in option '{}' in task {} was disabled for clashing with another option of same name", it.getName(), task.getPath()));
```
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/execution/commandline/CommandLineTaskConfigurer.java:66]: :art: This should be moved to `TaskOptionSupplier` so it is close to the similar logic for clashing opposite options.

Then we don't need non valid options in `CommandLineTaskConfigurer` this will simplify things.

Moreoever, the `help` task now warns about clashing opposites but not about clashing built-ins. Fixing this would make this uniform.

File: subprojects/core/src/main/java/org/gradle/execution/commandline/CommandLineTaskConfigurer.java [hunk 15]
```diff
@@ -94,4 +98,18 @@ private List<String> configureTasksNow(Collection<Task> tasks, List<String> argu
         }
         return remainingArguments;
     }
+
+    private void mutuallyExcludeOppositeOptions(CommandLineParser parser, List<OptionDescriptor> validCommandLineOptions) {
```
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/execution/commandline/CommandLineTaskConfigurer.java:102]: :art: `BooleanOptionElement` should not reference it's opposite.  IIUC this function is the only consumer of that.  `CommandLineTaskConfigurer` should not know how to calculate mutual exclusivity between boolean opposites. What I would do instead is have `TaskOptionSupplier` provide the right data ready for `CommandLineTaskConfigurer` to declare `allowOneOf` without doing any calculations. Then the whole concept of boolean opposites is encapsulated in `TaskOptionSupplier`.

This will further simplify things.

**Review by eskatos — CHANGES_REQUESTED** [review]
This is going into the right direction but should be taken a bit further in order to isolate complexity, simplify the command line task configurer and make the `help` task for uniform with it.

Also there are still many tests failing

**Comment by FrauBoes** [comment]
@bot-gradle test this

File: subprojects/core/src/main/java/org/gradle/execution/commandline/CommandLineTaskConfigurer.java [hunk 16]
```diff
@@ -94,4 +87,10 @@ private List<String> configureTasksNow(Collection<Task> tasks, List<String> argu
         }
         return remainingArguments;
     }
+
+    private void allowOneOfEach(CommandLineParser parser, List<OptionDescriptor> oppositeOptions) {
+        for (int i = 0; i < oppositeOptions.size(); i = i + 2) {
+            parser.allowOneOf(oppositeOptions.get(i).getName(), oppositeOptions.get(i+1).getName());
```
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/execution/commandline/CommandLineTaskConfigurer.java:93]: 🎨 It would be more readable to use a `List<Pair>` here instead of fiddling with the iteration index on a structure created in another object.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/TaskOptionSupplier.java [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+/**
+ * This class is responsible for adding built-in task options and
+ * {@link BooleanOptionElement boolean opposite options} to the options
+ * provided by the {@link OptionReader}.
+ */
+public class TaskOptionSupplier {
+    private static final Logger LOGGER = LoggerFactory.getLogger(TaskOptionSupplier.class);
+    private final Object target;
+    private final OptionReader optionReader;
+    private final List<OptionDescriptor> oppositeOptionPairs = new LinkedList<>();
```
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/TaskOptionSupplier.java:50]: 🎨 This new state here makes this object non-reusable and would cause suprise if it is reused. As we discussed I think instead of the two public functions which require holding this state we should have a single public function that returns an object that contains both all-options and mutually-exclusive-options.

**Comment by FrauBoes** [comment]
@bot-gradle test this

File: subprojects/platform-native/src/test/groovy/org/gradle/nativeplatform/toolchain/plugins/NativeToolChainPluginTest.groovy [hunk 18]
```diff
@@ -43,7 +43,7 @@ abstract class NativeToolChainPluginTest extends AbstractProjectBuilderSpec {
     }
 
     NativeToolChainInternal getToolchain() {
-        registry.get("toolChains", NativeToolChainRegistryInternal).getByName(getToolchainName()) as NativeToolChainInternal
+        registry.getAllOptions("toolChains", NativeToolChainRegistryInternal).getByName(getToolchainName()) as NativeToolChainInternal
```
Comment (eskatos) [inline subprojects/platform-native/src/test/groovy/org/gradle/nativeplatform/toolchain/plugins/NativeToolChainPluginTest.groovy:46]: ❓ This looks like an unrelated change

**Review by eskatos — CHANGES_REQUESTED** [review]
The simplification of `CommandLineTaskConfigurer` is very welcome 👍

File: subprojects/platform-native/src/test/groovy/org/gradle/nativeplatform/toolchain/plugins/NativeToolChainPluginTest.groovy (hunk 18 above)
Comment (FrauBoes) [reply subprojects/platform-native/src/test/groovy/org/gradle/nativeplatform/toolchain/plugins/NativeToolChainPluginTest.groovy:46]: Thanks for catching!

**Comment by FrauBoes** [comment]
@bot-gradle test this

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc [hunk 19]
```diff
@@ -387,6 +387,12 @@ beforeSettings {
 Using relative file paths to point to Java executables is now deprecated and will become an error in Gradle 9.
 This is done to reduce confusion about what such relative paths should resolve against.
 
+[[test_task_success_without_test]]
+==== Deprecated running the test task successfully without a test present
+Running the `Test` task successfully without a test present is now deprecated and will become an error in Gradle 9.
+This is done to avoid accidental successful test runs due to erroneous configuration.
+The behaviour can be controlled with the task options `--success-without-test` and `--no-success-without-test` or on the `Test` task using link:{javadocPath}/org/gradle/api/tasks/testing/Test.html#setSuccessWithoutTest-java.lang.Boolean-[setSuccessWithoutTest(Boolean)].
```
Comment (eskatos) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc:394]: :x: let's link to the `Property<Boolean> getSuccessWithoutTest()` lazy property instead of the command line option setter here.

**Comment by FrauBoes** [comment]
@bot-gradle test this

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java [hunk 20]
```diff
@@ -481,11 +485,25 @@ public void execute(DeleteSpec spec) {
...
-        } else if (testCountLogger.getTotalTests() == 0 && shouldFailOnNoMatchingTests()) {
-            throw new TestExecutionException(createNoMatchingTestErrorMessage());
+        } else if (testCountLogger.getTotalTests() == 0) {
+            if (shouldFailOnNoMatchingTests()) {
+                throw new TestExecutionException(createNoMatchingTestErrorMessage());
+            } else if (!successWithoutTest.isPresent()) {
+                emitDeprecationMessage();
+            } else if (!successWithoutTest.get()) {
+                throw new TestExecutionException(createNoMatchingTestErrorMessage());
+            }
```
Comment (marcphilipp) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java:495]: ❌ For Predictive Test Selection, we'll need a way to avoid failing here if there were > 0 tests discovered but 0 were selected.

**Review by marcphilipp — CHANGES_REQUESTED** [review]

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java (hunk 20 above)
Comment (eskatos) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java:495]: Thanks Marc
How could we achieve this in the best way for you?

Comment (marcphilipp) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java:495]: I think the `TestExecuter` would need a way to signal the number of tests it discovered, probably via a new method on `TestResultProcessor`. I've already reached out to @FrauBoes internally and we agreed to discuss options when she picks up work on this PR again.

Comment (FrauBoes) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java:495]: @marcphilipp Please see #24786 for work on the test task option. Note that I propose not to deprecate the current behaviour so it should not affect Predictive Test Selection anymore.

**Comment by FrauBoes** [comment]
This draft PR is closed now, please see the two resulting PRs:
- #24664 (merged)
- #24786 (in review)
